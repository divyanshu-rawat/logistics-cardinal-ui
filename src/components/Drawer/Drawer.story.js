import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withInfo } from '@storybook/addon-info';
import { color, boolean } from '@storybook/addon-knobs';

import withTests from '../../utils/withTests';

import { COMPONENTS } from '../../../.storybook/sections';

import Button from '../Button';
import ComposedTabs from '../Tabs';
import IDCard from '../IDCard';
import Text from '../Text';
import Drawer from '.';

const Header = <IDCard name="Teri Mitchell" info="Rider Captain" />;

const HorizontalTab = (
  <div style={{ position: 'relative', width: '700px', height: '100%' }}>
    <ComposedTabs
      mode="vertical"
      options={{
        forecast: (
          <Text as="p">
            Lorem Ipsum is simply dummy text of the printing and typesetting
            industry. Lorem Ipsum has been the industry's standard dummy text
            ever since the 1500s, when an unknown printer took a galley of type
            and scrambled it to make a type specimen book. It has survived not
            only five centuries, but also the leap into electronic typesetting,
            remaining essentially unchanged. It was popularised in the 1960s
            with the release of Letraset sheets containing Lorem Ipsum passages,
            and more recently with desktop publishing software like Aldus
            PageMaker including versions of Lorem Ipsum.
          </Text>
        ),
        history: (
          <Text as="p">
            There are many variations of passages of Lorem Ipsum available, but
            the majority have suffered alteration in some form, by injected
            humour, or randomised words which don't look even slightly
            believable. If you are going to use a passage of Lorem Ipsum, you
            need to be sure there isn't anything embarrassing hidden in the
            middle of text. All the Lorem Ipsum generators on the Internet tend
            to repeat predefined chunks as necessary, making this the first true
            generator on the Internet. It uses a dictionary of over 200 Latin
            words, combined with a handful of model sentence structures, to
            generate Lorem Ipsum which looks reasonable. The generated Lorem
            Ipsum is therefore always free from repetition, injected humour, or
            non-characteristic words etc.
          </Text>
        ),
      }}
    />
  </div>
);

storiesOf(`${COMPONENTS}Drawer`, module)
  .addDecorator(withTests('Drawer/tests/Drawer'))
  .add(
    'Simple Controlled by State',
    withInfo()(() => (
      <div style={{ width: '100%', height: '100%' }}>
        <Drawer
          header={Header}
          open={boolean('Open?', false)}
          right={boolean('Right?', true)}
          content={HorizontalTab}
        />
      </div>
    )),
  )
  .add(
    'With Render Props',
    withInfo()(() => (
      <div style={{ width: '100%', height: '100%' }}>
        <Drawer
          header={Header}
          open={boolean('Open?', false)}
          right={boolean('Right?', true)}
          content={HorizontalTab}
        >
          {({ isOpen }, toggleDrawer) => (
            <Button onClick={toggleDrawer}>Click me!</Button>
          )}
        </Drawer>
      </div>
    )),
  );
