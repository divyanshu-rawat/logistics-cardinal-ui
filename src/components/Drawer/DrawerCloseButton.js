import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Icons from '../Icons';

export const DrawerCloseButtonStyled = styled('a')`
  cursor: pointer;
`;

const DrawerCloseButton = ({ onClose }) => (
  <DrawerCloseButtonStyled onClick={onClose}>
    <Icons.Close className="Close" />
  </DrawerCloseButtonStyled>
);

DrawerCloseButton.propTypes = {
  onClose: PropTypes.func.isRequired,
};

export default DrawerCloseButton;
