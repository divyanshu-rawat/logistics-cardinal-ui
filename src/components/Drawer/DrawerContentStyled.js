import styled from 'styled-components';

const DrawerContentStyled = styled('div')`
  height: 100%;
`;

export default DrawerContentStyled;
