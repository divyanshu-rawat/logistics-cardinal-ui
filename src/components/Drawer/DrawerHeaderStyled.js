import styled from 'styled-components';

const DrawerHeaderStyled = styled('div')`
  min-height: 100px;
  display: flex;
  align-items: center;
  padding: ${({ theme }) => theme.composedSpacings.byte};
`;

export default DrawerHeaderStyled;
