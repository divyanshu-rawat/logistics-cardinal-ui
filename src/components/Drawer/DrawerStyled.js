import styled from 'styled-components';

const DrawerStyled = styled('div')`
  height: 100%;
  min-width: ${({ width }) => width}px;
  box-shadow: ${({ theme }) => theme.boxShadow.drawer};
  background-color: ${({ theme }) => theme.colors.grey300};
`;

export default DrawerStyled;
