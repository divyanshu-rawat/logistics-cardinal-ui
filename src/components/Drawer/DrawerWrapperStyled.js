import styled from 'styled-components';

const DrawerStyled = styled('div')`
  position: fixed;
  top: 0;
  width: auto;
  height: 100%;
  outline: 0;
  z-index: ${({ theme }) => theme.zIndex.drawerContent};
  ${({ right }) => {
    if (right) {
      return `right: 0;`;
    }
    return `left: 0;`;
  }};
  > .react-reveal {
    position: relative;
    height: 100%;
  }
`;

export default DrawerStyled;
