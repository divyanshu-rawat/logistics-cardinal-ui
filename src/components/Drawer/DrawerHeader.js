import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import DrawerHeaderStyled from './DrawerHeaderStyled';
import DrawerCloseButton from './DrawerCloseButton';

export const DrawerHeaderElement = styled('div')`
  flex: 1;
`;

export const DrawerHeaderButtonStyled = styled('div')`
  width: 50px;
  text-align: center;
`;

const DrawerHeader = ({ header, onClose }) => (
  <DrawerHeaderStyled>
    <DrawerHeaderElement>{header}</DrawerHeaderElement>
    <DrawerHeaderButtonStyled>
      <DrawerCloseButton onClose={onClose} />
    </DrawerHeaderButtonStyled>
  </DrawerHeaderStyled>
);

DrawerHeader.propTypes = {
  onClose: PropTypes.func.isRequired,
  header: PropTypes.node.isRequired,
};

export default DrawerHeader;
