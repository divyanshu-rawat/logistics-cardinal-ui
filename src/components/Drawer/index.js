import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { Fade } from 'react-reveal';

import DrawerStyled from './DrawerStyled';
import OverlayStyled from './OverlayStyled';
import DrawerWrapperStyled from './DrawerWrapperStyled';
import DrawerHeader from './DrawerHeader';
import DrawerContentStyled from './DrawerContentStyled';

export const BODY_OPEN_CLASSNAME = 'Drawer--is-open';

/**
 * @typedef {Object} Props
 * @property {boolean} open
 * @property {React.ReactNode} content
 * @property {boolean} right
 * @property {number} width
 * @property {boolean} [unselectDrawerWhen]
 * @property {function} [onToggle]
 * @property {function} [children]
 * @property {React.ReactNode} [header]
 *
 * @extends {Component<Props>}
 */
class Drawer extends Component {
  static propTypes = {
    open: PropTypes.bool.isRequired,
    content: PropTypes.node.isRequired,
    right: PropTypes.bool.isRequired,
    width: PropTypes.number.isRequired,
    unselectDrawerWhen: PropTypes.bool,
    onToggle: PropTypes.func,
    children: PropTypes.func,
    header: PropTypes.node,
  };

  static defaultProps = {
    open: false,
    right: false,
    unselectDrawerWhen: false,
    width: 300,
  };

  static DURATION = 500;
  static DELAY = 100;

  state = {
    isOpen: this.props.open || false,
  };

  componentDidMount() {
    this.bindEscKey();
  }

  componentDidUpdate(prevProps) {
    const { open, unselectDrawerWhen } = this.props;

    if (prevProps.open !== open && open !== this.state.isOpen) {
      this.toggleState();
    }

    if (open && prevProps.unselectDrawerWhen !== unselectDrawerWhen) {
      this.handleDrawerDepth();
    }
  }

  componentWillUnmount() {
    this.unbindEscKey();
  }

  onCloseEscape = (event) => {
    if (event.keyCode === 27) this.toggleState();
  };

  onOverlayClick = () => {
    this.toggleState();
  };

  handleCSSClasses = () => {
    const { isOpen } = this.state;

    if (isOpen) {
      document.body.classList.add(BODY_OPEN_CLASSNAME);
    } else {
      document.body.classList.remove(BODY_OPEN_CLASSNAME);
    }
  };

  toggleState = () => {
    const { onToggle } = this.props;
    this.setState({ isOpen: !this.state.isOpen }, () => {
      this.handleCSSClasses();

      if (!this.state.isOpen) {
        this.unbindEscKey();
      } else {
        this.bindEscKey();
      }

      if (onToggle) {
        onToggle(this.state.isOpen);
      }
    });
  };

  handleDrawerDepth = () => {
    const { unselectDrawerWhen } = this.props;

    if (unselectDrawerWhen) {
      return this.unbindEscKey();
    }

    return this.bindEscKey();
  };

  bindEscKey = () => {
    document.body.addEventListener('keyup', this.onCloseEscape);
  };

  unbindEscKey = () => {
    document.body.removeEventListener('keyup', this.onCloseEscape);
  };

  render() {
    const { children, right, content, header, width } = this.props;
    const { isOpen } = this.state;

    return (
      <Fragment>
        <DrawerWrapperStyled right={right}>
          <Fade
            when={isOpen}
            duration={Drawer.DURATION}
            delay={Drawer.DELAY}
            mountOnEnter
            unmountOnExit
            {...{ [right ? 'right' : 'left']: true }}
            style={{ position: 'relative' }}
          >
            <DrawerStyled width={width}>
              {header && (
                <DrawerHeader header={header} onClose={this.toggleState} />
              )}
              <DrawerContentStyled>{content}</DrawerContentStyled>
            </DrawerStyled>
          </Fade>
        </DrawerWrapperStyled>
        <OverlayStyled open={isOpen} onClick={this.onOverlayClick} />
        {children && children(this.state, this.toggleState)}
      </Fragment>
    );
  }
}

export default Drawer;
