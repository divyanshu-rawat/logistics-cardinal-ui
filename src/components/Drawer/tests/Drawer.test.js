import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Drawer from '..';

const renderedComponent = ({
  open = false,
  content = <p>Test!</p>,
  right = false,
  width = 500,
  header = <header>Header Test</header>,
  withRenderProps = false,
}) =>
  shallow(
    <Drawer
      open={open}
      content={content}
      right={right}
      width={width}
      header={header}
    >
      {withRenderProps // eslint-disable-next-line
        ? ({ isOpen }, toggleState) => (
          <button onClick={toggleState}>{isOpen}</button>
          )
        : null}
    </Drawer>,
  );

describe('<Drawer />', () => {
  it('should render the component', () => {
    const tree = toJson(renderedComponent({}));
    expect(tree).toMatchSnapshot();
  });

  it('should render the component width different props', () => {
    const tree = toJson(
      renderedComponent({ open: true, right: true, width: 100 }),
    );
    expect(tree).toMatchSnapshot();
  });

  it('should render the component with render props', () => {
    const tree = toJson(
      renderedComponent({
        open: true,
        right: true,
        width: 100,
        withRenderProps: true,
      }),
    );
    expect(tree).toMatchSnapshot();
  });

  it('should toggle state when overlay is clicked', () => {
    const component = renderedComponent({});
    component.setState({ isOpen: true });
    component.find('OverlayStyled').simulate('click');
    expect(component.state('isOpen')).toBe(false);
  });

  it('should keep the drawer opened if user is interacting with it', () => {
    const component = renderedComponent({});
    component.setState({ isOpen: true });
    component.find('DrawerStyled').simulate('click');
    expect(component.state('isOpen')).toBe(true);
  });

  it('should close the drawer when the ESC has been pressed', () => {
    const component = renderedComponent({});
    component.setState({ isOpen: true });
    const event = new KeyboardEvent('keyup', {
      key: 'Esc',
      which: 27,
      keyCode: 27,
    });
    document.body.dispatchEvent(event);
    expect(component.state('isOpen')).toBe(false);
    // if ESC key is pressed again it shouldn't call toggleState()
    document.body.dispatchEvent(event);
    expect(component.state('isOpen')).toBe(false);
  });

  it('should trigger toggleState when component has been updated', () => {
    const component = renderedComponent({});
    component.setProps({ open: true });
    expect(component.state('isOpen')).toBe(true);
  });

  it('should trigger toggleState when component has been updated', () => {
    const component = renderedComponent({});
    component.setState({ isOpen: true });
    const unbindEscKeyFn = jest.fn();
    component.instance().unbindEscKey = unbindEscKeyFn;
    component.unmount();
    expect(unbindEscKeyFn.mock.calls.length).toBe(1);
  });
});
