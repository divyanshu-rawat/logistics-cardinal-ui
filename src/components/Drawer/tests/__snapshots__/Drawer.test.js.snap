// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> should render the component 1`] = `
<Fragment>
  <DrawerWrapperStyled__DrawerStyled
    right={false}
  >
    <Fade
      delay={100}
      duration={500}
      left={true}
      mountOnEnter={true}
      style={
        Object {
          "position": "relative",
        }
      }
      unmountOnExit={true}
      when={false}
    >
      <DrawerStyled
        width={500}
      >
        <DrawerHeader
          header={
            <header>
              Header Test
            </header>
          }
          onClose={[Function]}
        />
        <DrawerContentStyled>
          <p>
            Test!
          </p>
        </DrawerContentStyled>
      </DrawerStyled>
    </Fade>
  </DrawerWrapperStyled__DrawerStyled>
  <OverlayStyled
    onClick={[Function]}
    open={false}
  />
  <button
    onClick={[Function]}
  />
</Fragment>
`;

exports[`<Drawer /> should render the component width different props 1`] = `
<Fragment>
  <DrawerWrapperStyled__DrawerStyled
    right={true}
  >
    <Fade
      delay={100}
      duration={500}
      mountOnEnter={true}
      right={true}
      style={
        Object {
          "position": "relative",
        }
      }
      unmountOnExit={true}
      when={true}
    >
      <DrawerStyled
        width={100}
      >
        <DrawerHeader
          header={
            <header>
              Header Test
            </header>
          }
          onClose={[Function]}
        />
        <DrawerContentStyled>
          <p>
            Test!
          </p>
        </DrawerContentStyled>
      </DrawerStyled>
    </Fade>
  </DrawerWrapperStyled__DrawerStyled>
  <OverlayStyled
    onClick={[Function]}
    open={true}
  />
  <button
    onClick={[Function]}
  >
    <Component />
  </button>
</Fragment>
`;
